---
title: "We are hiring!"
authors:
- phillippy
---
The Genome Informatics Section is hiring! Come and join our outstanding team at the [National Human Genome Research Institute](https://genome.gov/dir). The official advertisement is for a postdoctoral position, but we will also consider outstanding postbacs or graduate students (via the [NIH GPP](https://www.training.nih.gov/programs/gpp)). Simply follow the same application procedure described below. **Update 2018-02-22: this position has now been filled.**

<!--excerpt-->

A postdoctoral research position is available under the supervision of [Dr. Adam Phillippy](https://genome.gov/staff/phillippy) at the [National Human Genome Research Institute](https://www.genome.gov/dir/) (NHGRI). Dr. Phillippy leads the Genome Informatics Section, which develops and applies computational methods for the analysis of massive genomics datasets with a focus on problems related to genome sequencing. Members of the section have developed many widely used bioinformatics methods and are highly collaborative, often tackling new problems with multiple partner laboratories. Following this tradition, the section continues to foster open and interdisciplinary collaboration between the computational, biological, and medical sciences for the advancement of global health.

The section is currently seeking applicants with an interest in developing and/or applying computational methods for genome assembly, sequence alignment, structural variant detection, metagenomics, information visualization, and machine learning. Our recent work includes the development of the Canu genome assembler ([Koren et al. 2017](https://doi.org/10.1101/gr.215087.116)), assembly of the first human genome from nanopore sequencing data ([Jain et al. 2017](https://doi.org/10.1101/128835)), and highly scalable comparative genomics tools, such as Mash ([Ondov et al. 2016](https://dx.doi.org/10.1186/s13059-016-0997-x)). Ongoing work includes efforts to finish the human reference genome, generate high-quality reference genomes for all vertebrate taxonomic orders, and develop real-time nanopore sequencing assays for the diagnosis of infectious disease, cancers, and genetic disorders.

Candidates should have or be close to obtaining a Ph.D. or equivalent degree in a field related to computer science, applied mathematics, bioinformatics, or genomics. Applicants must possess strong programming, English, and analytical skills. *The NIH is dedicated to building a diverse community in its training and employment programs.*

The Intramural Research Program is on the NIH Bethesda campus and offers a wide array of training and collaboration opportunities for early-career scientists. The funding for this position is stable and offers the trainee wide latitude in the design and pursuit of their own research. The successful candidate will have access to extensive high-performance computing resources ([BioWulf](https://hpc.nih.gov/)), the NIH intramural sequencing center ([NISC](https://www.nisc.nih.gov/)), NHGRI core facilities, and collaboration with the NIH Clinical Center. A typical stipend for a computational postdoc is between $60â€“70k per year and includes family health insurance. Answers to some frequently asked questions can be found here: [NIH Postdoc FAQs](https://www.training.nih.gov/resources/faqs/postdoc_irp).

**To apply:** Interested applicants should submit their CV, a brief research statement, and the names of three references to: adam.phillippy@nih.gov
